### PR TITLE
Update XpressionMapper to handle ForPath mapper configurations.

### LINKED ISSUE #2293: Reopen issue #2285 Can expression mapping work with ForPath?
Reopening this issue to make unflattening simpler [see the question on stack overflow](https://stackoverflow.com/questions/45802903/automapper-expression-mapping).

### PR DESCRIPTION
**Pull request by BlaiseD**
Fixes #2293.

### CODE REVIEW

File: src/AutoMapper/XpressionMapper/FindMemberExpressionsVisitor.cs [hunk 1]
```diff
@@ -44,19 +44,16 @@ public MemberExpression Result
...
-                                                                                    || node.Type.GetTypeInfo().IsValueType
-                                                                                    || (node.Type.GetTypeInfo().IsGenericType 
-                                                                                        && node.Type.GetGenericTypeDefinition() == typeof(Nullable<>)
-                                                                                        && Nullable.GetUnderlyingType(node.Type).GetTypeInfo().IsValueType)))
-                    _memberExpressions.Add((MemberExpression)node.Expression);
+                if (node.Expression.NodeType == ExpressionType.MemberAccess && (node.Type == typeof(string)
+                                                                                    || node.Type.GetTypeInfo().IsValueType))
+                    _memberExpressions.Add((MemberExpression)node.Expression);
+                else if (node.Expression.NodeType == ExpressionType.Parameter && (node.Type == typeof(string)
+                                                                                    || node.Type.GetTypeInfo().IsValueType))
```
Comment (lbargaoanu) [inline src/AutoMapper/XpressionMapper/FindMemberExpressionsVisitor.cs]: I think a little extension method encapsulating this test (string or value type) might make the code more readable. Of course I don't know much about expression mapping, just a thought :) A good name goes a long way.

File: src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs [hunk 2]
```diff
@@ -231,9 +231,16 @@ protected void FindDestinationFullName(Type typeSource, Type typeDestination, st
             if (typeMap == null)
             {
                 throw QueryMapperHelper.MissingMapException(sourceType: typeDestination, destinationType: typeSource);
+            }
+
+            PathMap pathMap = typeMap.PathMaps.SingleOrDefault(item => string.Join(period, item.MemberPath.Members.Select(m => m.Name)) == sourceFullName);
```
Comment (lbargaoanu) [inline src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs]: Maybe make this a TypeMap method?

File: src/UnitTests/XpressionMapper.ForPath.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                .ForPath(d => d.Nested.NestedTitle, opt => opt.MapFrom(src => src.Title))
+                .ForPath(d => d.Nested.NestedTitle2, opt => opt.MapFrom(src => src.Title2));
+
+            CreateMap<OrderDto, Order>()
+                .ForPath(o => o.CustomerHolder.Customer.Name, o => o.MapFrom(s => s.Customer.Name))
+                .ForPath(o => o.CustomerHolder.Customer.Total, o => o.MapFrom(s => s.Customer.Total))
+                .ForPath(o => o.CustomerHolder.Customer.Address, o => o.MapFrom(s => s.CustomerAddress))
+                .ForPath(o => o.CustomerHolder.Customer.Age, o => o.MapFrom(s => s.CustomerAge));
+
+            CreateMissingTypeMaps = true;
```
Comment (lbargaoanu) [inline src/UnitTests/XpressionMapper.ForPath.Tests.cs]: Please no :)

File: src/AutoMapper/XpressionMapper/FindMemberExpressionsVisitor.cs (hunk 1 above)
Comment (BlaiseD) [reply src/AutoMapper/XpressionMapper/FindMemberExpressionsVisitor.cs]: Makes sense  - done.

File: src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs (hunk 2 above)
Comment (BlaiseD) [reply src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs]: Less value here I think but no downside - done.

File: src/UnitTests/XpressionMapper.ForPath.Tests.cs (hunk 3 above)
Comment (BlaiseD) [reply src/UnitTests/XpressionMapper.ForPath.Tests.cs]: Made it nullable :)

Comment (lbargaoanu) [reply src/UnitTests/XpressionMapper.ForPath.Tests.cs]: OK, I should have said what I meant :) I have an issue with setting CreateMissingTypeMaps.

File: src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs (hunk 2 above)
Comment (lbargaoanu) [reply src/AutoMapper/XpressionMapper/XpressionMapperVisitor.cs]: Yes, probably no one else will use it, but it's no biggie.

File: src/UnitTests/XpressionMapper.ForPath.Tests.cs (hunk 3 above)
Comment (BlaiseD) [reply src/UnitTests/XpressionMapper.ForPath.Tests.cs]: Ah -  got it.

**Comment by BlaiseD** [comment]
No - looks good.  Thanks.

**Pull request merged**